### PR TITLE
Add new SearchButton to all pages and tweak SearchRouter

### PR DESCRIPTION
**Pull request by Kicu**
### Details
 - This PR both implements issue xxx and also introduces some followups after merging SearchRouter (https://github.com/Expensify/App/pull/48785)
 - `SearchButton` is added to almost every page (as listed in the design doc)
 - there is one exception: rooms list on mobile - because we still have the ChatFinder there, so this time I'm avoiding two search icons; `ChatFinder` will be removed in a separate issue
 - I had to add 1 prop to SelectionList to pass it to `useKeyboardShortcut` hook, since you can trigger search from router via Enter key, but `useKeyboardShortcut` consumed the ENTER keypress event

_misc notes:_
 - [tweaks] standardised naming of some SearchUtils functions; dropped one function that was only used in once place
 - `WorkspacesListPage` and `TroubleShootPage` were migrated to `useOnyx` in compliance with new CI check

### Fixed Issues

$ https://github.com/Expensify/App/issues/49122
$ https://github.com/Expensify/App/issues/49118
PROPOSAL:

### Tests

- verify that the button to trigger new search is displayed on every page
- verify that clicking the button opens new search router
- verify that search router works - it should navigate to search with specific query

### QA Steps

- verify that the button to trigger new search is displayed on every page
- verify that clicking the button opens new search router
- verify that search router works - it should navigate to search with specific query

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/886ee00f-5388-41ca-8236-4abe2d0c6a41

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/c81f10eb-0c33-4526-85e5-d7550e8957b2

</details>

<details>
<summary>iOS: mWeb Safari</summary>
https://github.com/user-attachments/assets/e9eb66ff-03af-4fde-a497-8770b7c6dff5

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/f935e898-0d16-440a-abf5-ece884df798c

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

File: src/libs/SearchUtils.ts [hunk 1]
```diff
@@ -407,6 +397,20 @@ function getQueryHashFromString(query: SearchQueryString): number {
     return UserUtils.hashText(query, 2 ** 32);
 }
 
+function getExpenseTypeTranslationKey(expenseType: ValueOf<typeof CONST.SEARCH.TRANSACTION_TYPE>): TranslationPaths {
```
Comment (Kicu) [inline src/libs/SearchUtils.ts:400]: only moved

File: src/libs/SearchUtils.ts [hunk 2]
```diff
@@ -467,16 +471,58 @@ function sanitizeString(str: string) {
...
-            return 'common.distance';
-        case CONST.SEARCH.TRANSACTION_TYPE.CARD:
-            return 'common.card';
-        case CONST.SEARCH.TRANSACTION_TYPE.CASH:
-            return 'iou.cash';
+/**
+ * @private
+ * traverses the AST and returns filters as a QueryFilters object
+ */
+function getFilters(queryJSON: SearchQueryJSON) {
```
Comment (Kicu) [inline src/libs/SearchUtils.ts:478]: only moved

**Comment by Kicu** [comment]
@luacmartins originally I planned for this PR to be something else, but now the scope changed a bit.
Ready for review

**Comment by ikevin127** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/bdaa55fb-15fc-42eb-83ac-6ba66f75e2df

</details>

<details>
<summary>Android: mWeb Chrome</summary>

Uploading android-mweb.mov…

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/7ac91ef7-a263-4613-90eb-d946567fbddb

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/22520a46-778d-4a2e-9b15-ac4cf10bd6af

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/3f99b1b7-cf4e-4cdd-9b91-4a6da03ab133

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/5b6925fc-f71c-48a0-a618-7544605aa348

</details>

**Comment by ikevin127** [comment]
> [!note]
> ### Observations
> 1. Missing  🔍  in Settings > `Subscription` page (probably on purpose ?).
> 2. When search input was performed and user is navigated to search page with specific query > 🔍 is missing from `Filters` button right side on wide layout devices (web / desktop).
> 3. (already present on staging) When search input was performed and user is navigated to search page with specific query  > LHN `Expenses` menu item is **not** highlighted.

Please confirm whether all of the above are expected.
Once addressed and the reassure tests are fixed, I'll be able to 🟢 **Approve**.

**Comment by luacmartins** [comment]
>Missing 🔍 in Settings > Subscription page (probably on purpose ?).

I think it should be present there too.

>When search input was performed and user is navigated to search page with specific query > 🔍 is missing from Filters button right side on wide layout devices (web / desktop).

I think this one is expected for now, since those pages will have a full width search bar instead of the magnifying glass icon

>(already present on staging) When search input was performed and user is navigated to search page with specific query > LHN Expenses menu item is not highlighted.

This is expected since at that point the user is seeing a custom search

**Comment by luacmartins** [comment]
@Expensify/design should the magnifying glass icon be to the left or right of existing header buttons? example:

Screenshot 2024-09-23 at 9 10 05 AM

File: src/components/HeaderWithBackButton/types.ts [hunk 3]
```diff
@@ -128,6 +128,9 @@ type HeaderWithBackButtonProps = Partial<ChildrenProps> & {
     /** Whether we should overlay the 3 dots menu */
     shouldOverlayDots?: boolean;
 
+    /** Whether we should display button that opens new SearchRouter */
```
Comment (luacmartins) [inline src/components/HeaderWithBackButton/types.ts:131]:
```suggestion
    /** Whether we should display the button that opens new SearchRouter */
```

**Comment by dubielzyk-expensify** [comment]
> @Expensify/design should the magnifying glass icon be to the left or right of existing header buttons? example:

Great question. I think we'll always want it on the right to ensure the router stays consistently in the top right corner when it opens.

**Comment by Kicu** [comment]
> I think we'll always want it on the right to ensure the router stays consistently in the top right corner when it opens.

Yeah I made a similar assumption and I'm always putting it furthest to the right on all the views.

**Comment by Kicu** [comment]
@ikevin127 @luacmartins icon added to Subscription page

Screenshot 2024-09-23 at 13 16 39

As for the other two like Carlos said they are expected. LHN items are only highglighted when you are on the exact query that the buttons link to, so `type:Expenses status:all` etc.

There is a separate PR to add full width input to Search Results page, and that is why we don't need 🔍 there. (See https://github.com/Expensify/App/pull/49462)

**Review by ikevin127 — APPROVED** [review]
🟢 This is looking good, I just verified that the search button was added to the Subscription page.

##

@Kicu ⚠️ The only thing left to do here is address the failing performance tests, which from what I can see the issue might come from the permissions check in `SearchButton` which does not seem to be mocked given the error message of the failing performance test:

```bash
● [ReportScreen] should render report list

    TypeError: _Permissions.default.canUseNewSearchRouter is not a function

      15 |     const {openSearchRouter} = useSearchRouterContext();
      16 |
    > 17 |     if (!Permissions.canUseNewSearchRouter()) {
         |                      ^
      18 |         return;
      19 |     }
      20 |
```

**Review by luacmartins — DISMISSED** [review]
LGTM

**Comment by luacmartins** [comment]
Unrelated jest tests are failing. Gonna see if there's a fix for it already

**Comment by Kicu** [comment]
Main merged and the missing mock in perf tests fixed, thanks for pointing this out @ikevin127 

Now the `Changed files ESLint check` is failing, but we can ignore this one. It fails on the deprecation of `measurePerformance` (because I touched the perf test file) which is not something we are handling right now.

If nothing else fails then I think we're good to go.

**Review by luacmartins — APPROVED** [review]

**Comment by luacmartins** [comment]
Failing tests are unrelated to the changes in this PR, merging.

**Pull request merged**